x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,284 @@
+        Path path = environment.configFile().resolve(keyStorePath);
+        KeyStore keyStore = readKeyStore(path, keystore_password);
+        final List<Tuple<PrivateKey, X509Certificate>> httpCaKeysAndCertificates =
+        getPrivateKeyEntries(keyStore, keystore_password).stream()
+            .filter(t -> t.v2().getBasicConstraints() != -1).collect(Collectors.toList());
+        if(httpCaKeysAndCertificates.isEmpty()) {
+            throw new IllegalStateException("Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration " +
+                "Keystore doesn't contain any PrivateKey entries where the associated certificate is a CA certificate");
+        } else if(httpCaKeysAndCertificates.size()>1) {
+            throw new IllegalStateException("Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration " +
+                "Keystore contain multiple PrivateKey entries where the associated certificate is a CA certificate");
+        }
+        return SslUtil.calculateFingerprint(httpCaKeysAndCertificates.get(0).v2());
+    }
+
+    static List<String> getFilteredAddresses(List<String> addresses) throws Exception {
+        List<String> filtered_addresses = new ArrayList<>();
+        for (String bound_address : addresses){
+            URI uri = new URI("http://" + bound_address);
+            InetAddress inet_address = InetAddress.getByName(uri.getHost());
+            if (inet_address.isLoopbackAddress() != true || inet_address.isSiteLocalAddress() == true) {
+                filtered_addresses.add(bound_address);
+            }
+        }
+        return filtered_addresses.isEmpty() ? addresses : filtered_addresses;
+    }
+
+    static KeyStore readKeyStore(Path path, SecureString password) throws GeneralSecurityException {
+        final String type = inferKeyStoreType(path);
+        if (Files.notExists(path)) {
+            throw new SslConfigException("cannot read a [" + type + "] keystore from [" + path.toAbsolutePath()
+                + "] because the file does not exist");
+        }
+        try {
+            KeyStore keyStore = KeyStore.getInstance(type);
+            try (InputStream in = Files.newInputStream(path)) {
+                keyStore.load(in, password.getChars());
+            }
+            return keyStore;
+        } catch (IOException e) {
+            throw new SslConfigException("cannot read a [" + type + "] keystore from [" + path.toAbsolutePath() + "] - " + e.getMessage(),
+                e);
+        }
+    }
+
+    static String inferKeyStoreType(Path path) {
+        String name = path == null ? "" : path.toString().toLowerCase(Locale.ROOT);
+        if (name.endsWith(".p12") || name.endsWith(".pfx") || name.endsWith(".pkcs12")) {
+            return "PKCS12";
+        } else {
+            return "jks";
+        }
+    }
+
+    static List<Tuple<PrivateKey, X509Certificate>> getPrivateKeyEntries(KeyStore keyStore, SecureString password) {
```

Comment:
can we use the method from KeyManager and have that be static instead of copying the code in here too ?

Reply:
You will have to pass the passwords and the path to call it, which as I understand wasn't a desirable behavior